### ISSUE TITLE
[FEAT]: Action trigger at battery percentage

### ISSUE THREAD

**Issue by Xerionius** [BartoszCichecki/LenovoLegionToolkit#564]
### Version

2.9.1

### OS

Windows 11

### Device

Legion 5 Pro 16ACH

### Is your feature request related to a problem?

No

### How would you like the problem to be solved?

Having a certain battery percentage as a trigger for actions would be great. You should be able to specify at which battery percentage the action should trigger and whether it should trigger on rising or falling battery level.
This way you could enable conservation mode on a certain battery percentage to get around the default limit of 60 %. And you could change the refresh rate, resolution and brightness when the battery gets low.

### What alternatives have you considered?

Instead of differentiating between rising and falling battery levels, it could also differentiate between "plugged in" and "not plugged in".

**Comment by BartoszCichecki** [comment]
This was discussed before, but can't find the issue now. Anyway, triggering Actions at certain battery percentage is not planned, because it can't be implemented reliably: for example actions won't run when laptop is sleeping or off (naturally).

I see you mention Conservation mode limits - long story short if the 60% limit is too low for you, it kind of indicates that you use your laptop on battery, so why do you use conservation mode? (Previously discussed here: https://github.com/BartoszCichecki/LenovoLegionToolkit/issues/20#issuecomment-988825583)

The alternative you presented is already implemented: there are actions for plugging in AC adapter.

Lastly the idea with adding "Low battery" trigger is good, let me know if you'd be interested in that. (ref: https://learn.microsoft.com/en-us/windows/win32/power/pbt-apmbatterylow)

**Comment by Xerionius** [comment]
I understand the drawbacks and for me it would still be useful. However I agree that it might not make sense for most users, unfortunately.

I use my laptop mostly on AC power but I would like to have it ready to go when I need it. At 60 % the battery just doesn't last long enough to be useful and I feel like 80 or 90 % would be a good compromise, as 90 % already decreases wear significantly compared to 100 %.

With the alternative, I was suggesting still making it dependent on battery percentage, but instead of differentiating between rising and falling, differentiating between battery and AC power, which would be the same thing in practice.

Personally I would not be that interested in a low battery trigger, because I already have set the trigger for AC power disconnection to take some power saving measures.

**Comment by BartoszCichecki** [comment]
OK, I will keep the issue open but on hold. I need to think if I can fit this into the current architecture without making major changes.

**Comment by vink2212** [comment]
I agree with @Xerionius I've wanted a feature like this for a while. Being able to trigger actions at certain battery %ages. A disclaimer that it won't work reliably with the machine sleeping would be good so people understand the limitations of the feature.

**Comment by bcutter** [comment]
Looking forward to use this too.

**Comment by Ekscentricitet** [comment]
I am interested in implementing the low battery trigger. I took a look into it and it would take me some time to go through the code base and understand what's happening and how I can integrate a listener for the event you linked. If you have time to sketch what changes have to be made I would give it a try.

**Comment by Snake-babu** [comment]
After the introduction of LLT CLI, I was able to code a fairly simple python script to achieve this, I will attach it to this comment
- It is a WIP I haven't even properly tested if it completely works 
- Honestly a shit solution, but good enough for me 
- This code was made for my personal use, but you can edit it for your own needs
- Like @BartoszCichecki mentioned it won't work if the laptop is in sleep/shut down. 

[charge test.txt](https://github.com/user-attachments/files/16250589/charge.test.txt)

**Comment by k3nsh1nta** [comment]
BC why long term, short term and peak power are limited. Is it difficult to add lower values?

**Comment by BartoszCichecki** [comment]
First and last warning for spamming @k3nsh1nta.

**Comment by k3nsh1nta** [comment]
where exactly did i spam inside here about cpu watt adjustment? if you can point me out the comments, then im really sorry.

**Comment by BartoszCichecki** [comment]
How is CPU wattage relevant to action triggers for battery percentage? Again, last warning. This time for real.

**Comment by k3nsh1nta** [comment]
Searched for power limit on search bar of issues in order to not creating or spamming this thing if its already had been asked etc, as you instructed me with the fan curve, and only this post appeared. If this considered as spam ( cause in my knowledge spamming is talking, typing or generally speaking the same thing over and over again ) you ve my sorries for disturbing you and you can delete the comments. I didnt ve the purpose to spam anything, my sorries again.

**Comment by bayard** [comment]
Fan speed normally needed to run higher and cause louder noise due to aging of thermal paste and accumulation of dust as time goes by. 

Fan will be much quitter when charging stopped, so trigger charging mode at certain battery percentage will be very helpful to reduce noise, like resume charging at 30% and stop charging at 90%.

Hope the feature will be added in future, or as a experimental feature for users who been annoyed by noise for long time. 

I always set the CPU run at 0w+ of PL1/PL2 and DynamicBoost disabled, but it slows the laptop and cause UI lagging at noticeable level, it will be much great if trigger feature introduced.

Thanks a lot!

**Comment by BartoszCichecki** [comment]
> Fan speed normally needed to run higher and cause louder noise due to aging of thermal paste and accumulation of dust as time goes by.

Fixing your thermals is beyond the scope of this project. Regarding your argument, there is only one solution: clean your fans.
